src/sagemaker_xgboost_container/distributed_gpu/dask_cluster_utils.py

```diff
@@ -0,0 +1,40 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License'). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the 'license' file accompanying this file. This file is
+# distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import socket
+from subprocess import Popen
+
+from sagemaker_algorithm_toolkit.exceptions import PlatformError
+
+SCHEDULER_EXEC_PATH = "/miniconda3/bin/dask-scheduler"
+CUDA_WORKER_EXEC_PATH = "/miniconda3/bin/dask-cuda-worker"
+
+
+def start_daemons_in_current_instance(scheduler_conn_string: str, is_scheduler_host: bool):
```

Comment:
I'm still a little confused about this function and why we use `Popen` instead of `call`. Below, we call this function and then wait `WAIT_TIME_DAEMONS_STARTUP_SEC`, presumably for the scheduler and worker to come online, right? I think if we used `call` here, which is blocking, this function wouldn't return until the scheduler and worker came online, which I think is exactly what we're trying to wait for with the sleep.

Reply:
Good point. My reasoning for using Popen and explicit wait seems to be flawed now. Call seems simpler. Will update.

Reply:
Actually, wouldn't call just block since the commands will be starting long running processes? Unless I'm understanding the documentation wrong it only exists once the command completes: https://docs.python.org/3/library/subprocess.html#subprocess.call

Reply:
Ah you're right - I misunderstood what those calls would return.

Popen should be correct here but we might want to think about whether there's a better way to check whether the scheduler and workers are online than a constant wait. Is there any way we can do a health check of sorts?

Reply:
I see that the [Client](https://distributed.dask.org/en/stable/api.html#distributed.Client) object has a timeout param. Not sure what the default timeout value is, but that param + wait_for_workers should be enough. Will test this out.

Reply:
No, that doesn't work. Without explicit wait some workers start up before the scheduler and these workers don't seem to retry connecting to the scheduler. So the cluster ends up with very few workers. 
The other option seems to be to create a Client to check if scheduler is up, before starting the workers. However, I'd like to backlog this. Lmk if there are concerns.